source/extensions/filters/http/proto_api_scrubber/filter_config.cc

```diff
@@ -349,43 +369,75 @@ void ProtoApiScrubberFilterConfig::initializeTypeUtils() {
       });
 }
 
-absl::StatusOr<const MethodDescriptor*>
-ProtoApiScrubberFilterConfig::getMethodDescriptor(const std::string& method_name) const {
-  // Covert grpc method name from `/package.service/method` format to `package.service.method` as
-  // the method `FindMethodByName` expects the method name to be in the latter format.
-  std::string dot_separated_method_name =
-      absl::StrReplaceAll(absl::StripPrefix(method_name, "/"), {{"/", "."}});
-  const MethodDescriptor* method = descriptor_pool_->FindMethodByName(dot_separated_method_name);
-  if (method == nullptr) {
-    return absl::InvalidArgumentError(absl::StrFormat(
-        "Unable to find method `%s` in the descriptor pool configured for this filter.",
-        dot_separated_method_name));
+void ProtoApiScrubberFilterConfig::resolveAndCacheType(
+    const std::string& raw_type_name, const std::string& package_prefix,
+    const std::string& method_key,
+    absl::flat_hash_map<std::string, const Envoy::Protobuf::Type*>& cache,
+    absl::string_view type_category) {
+  std::string type_name = raw_type_name;
+
+  // Handle fully qualified names starting with "." or append package prefix
+  if (absl::StartsWith(type_name, ".")) {
+    type_name = type_name.substr(1);
+  } else if (!package_prefix.empty()) {
+    type_name = absl::StrCat(package_prefix, type_name);
   }
 
-  return method;
+  std::string type_url = absl::StrCat(Envoy::Grpc::Common::typeUrlPrefix(), "/", type_name);
+
+  if (const auto* type_ptr = (*type_finder_)(type_url)) {
+    cache[method_key] = type_ptr;
+  } else {
+    ENVOY_LOG(error, "Failed to resolve {} Type for {}. URL: {}", type_category, method_key,
+              type_url);
+  }
+}
+
+void ProtoApiScrubberFilterConfig::precomputeTypeCache(
+    const Envoy::Protobuf::FileDescriptorSet& descriptor_set) {
+  for (const auto& file : descriptor_set.file()) {
+    std::string package_prefix;
+    if (!file.package().empty()) {
+      package_prefix = absl::StrCat(file.package(), ".");
+    }
+
+    for (const auto& service : file.service()) {
+      for (const auto& method : service.method()) {
+        // Construct the Method Key (e.g., /package.Service/Method).
+        std::string method_key =
+            absl::StrCat("/", package_prefix, service.name(), "/", method.name());
+
+        resolveAndCacheType(method.input_type(), package_prefix, method_key, request_type_cache_,
```

Comment:
It may be better to ensure that an entry isn't already cached, but this can be done as part of future work.